### PR TITLE
RSDK-2719 - Namespace all the things

### PR DESCRIPTION
**Pull request by stuqdog**
#### Major changes
- Add consistent `namespace`ing to the whole repo.

### CODE REVIEW

File: src/tests/mocks/camera_mocks.hpp [hunk 1]
```diff
@@ -8,6 +8,9 @@
 #include <components/camera/client.hpp>
 #include <components/camera/server.hpp>
 
+namespace viam {
+namespace cppsdk {
```
Comment (zaporter-work) [inline src/tests/mocks/camera_mocks.hpp:12]: Can we namespace this one layer deeper? Like viam::cppsdk::cameramocks? Otherwise functions like `fake_...` can collide when compiling the tests. (I ran into this when adding additional motor tests.)

**Review by zaporter-work — APPROVED** [review]
This is beautiful. I had been thinking about doing this myself.

File: examples/dial/example_dial.cpp [hunk 2]
```diff
@@ -22,16 +22,15 @@
 #include <rpc/dial.hpp>
 
 using viam::robot::v1::Status;
-using Viam::SDK::Credentials;
-using Viam::SDK::Options;
+using namespace viam::cppsdk;
```
Comment (acmorrow) [inline examples/dial/example_dial.cpp:25]: This is fine in an example, but see below for how it should be done in SDK source files.

Comment (acmorrow) [inline examples/dial/example_dial.cpp:25]: I want to think a bit about whether `viam` is right at the top-level (I think so), and whether `cppsdk` is the right inner name (I suspect it is).

File: src/tests/mocks/camera_mocks.hpp (hunk 1 above)
Comment (acmorrow) [reply src/tests/mocks/camera_mocks.hpp:12]: A case can be made that the mocks or other test code doesn't actually belong in the `viam:cppsdk` namespace at all.

I'm also not usually too keen on a complex set of "inner namespaces" due to the fact that namespaces have interactions with name lookup: https://en.cppreference.com/w/cpp/language/adl

**Review by acmorrow — CHANGES_REQUESTED** [review]

**Comment by acmorrow** [comment]
> This is beautiful. I had been thinking about doing this myself.

Agreed I'm very happy to see this happening.

File: examples/modules/CMakeLists.txt [hunk 3]
```diff
@@ -17,6 +17,7 @@ add_executable(example_module
 )
 
 target_link_libraries(example_module
+  PRIVATE Threads::Threads
```
Comment (stuqdog) [inline examples/modules/CMakeLists.txt:20]: (driveby) make sure we link threads properly. this was necessary to fix builds on linux.

File: src/tests/mocks/camera_mocks.cpp [hunk 4]
```diff
@@ -8,6 +8,9 @@
 #include <components/camera/server.hpp>
 #include <tests/test_utils.hpp>
 
+namespace viam {
+namespace cppsdktests {
```
Comment (zaporter-work) [inline src/tests/mocks/camera_mocks.cpp:12]: This still has the problem of declaring functions like `fake_properties` in the namespace for all of the tests but other tests (like motor) will also have `fake_properties` (if it follows the same pattern). Thoughts?

**Review by zaporter-work — APPROVED** [review]

File: src/tests/mocks/camera_mocks.hpp (hunk 1 above)
Comment (zaporter-work) [reply src/tests/mocks/camera_mocks.hpp]: This still has the problem of declaring functions like `fake_properties` in the namespace for all of the tests but other tests (like motor) will also have `fake_properties` (if it follows the same pattern). Thoughts?

Comment (stuqdog) [reply src/tests/mocks/camera_mocks.hpp]: Ah, good catch. I added an additional layer of namespacing for the `mock`s only, and using the appropriate namespace in associated tests. This should keep generic functions like `fake_properties` separate, and hopefully minimize namespace bloat. @acmorrow given that this is confined to tests my intuition is it's probably okay as far as the "inner namespace" issue goes but I'm curious to your thoughts.

File: src/components/service_base.cpp [hunk 5]
```diff
@@ -1 +1,4 @@
 #include <components/service_base.hpp>
+
+namespace viam {
+namespace cppsdk {
```
Comment (acmorrow) [inline src/components/service_base.cpp:4]: Missing closing namespace `}`'s here. Is this file actually registered in the build system?

File: src/resource/resource.hpp [hunk 6]
```diff
@@ -102,6 +105,10 @@ class Model {
     friend bool operator==(const Model& lhs, const Model& rhs);
 };
 
+}  // namespace cppsdk
+}  // namespace viam
+
+using namespace viam::cppsdk;
```
[the remark this answers is not in the document]
Comment (acmorrow) [reply src/resource/resource.hpp:111]: No, you can't do `using namespace viam::cppsdk` in a header, because it defeats the namespacing for consumers (direct or indirect) of that header. Instead, I'd expect you can just fully qualify our types:

```
template<>
struct std::hash<::viam::cppsdk::Name> {
...
```

Please give that a try and LMK if it doesn't work out as you expect for some reason.

**Review by acmorrow — CHANGES_REQUESTED** [review]
Looking good. I think this is ready to go mod the hash customization.

File: src/components/service_base.cpp (hunk 5 above)
Comment (stuqdog) [reply src/components/service_base.cpp:4]: Thanks for the catch, fixed. It looks like we have the header in the `CmakeLists.txt` but not the implementation file (which is probably fine since it's empty). Would rather not touch in this PR, but I actually already made a ticket to look into this file (which I suspect we can get rid of entirely with some light refactoring) [here](https://viam.atlassian.net/browse/RSDK-2718).

File: src/resource/resource.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (stuqdog) [reply src/resource/resource.hpp]: That works! Change made.

**Review by acmorrow — APPROVED** [review]
LGTM. Ship it! If I have some sort of naming epiphany it is a sed away given what is now in place.
